apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java

```diff
@@ -384,11 +451,29 @@ private void fillTraceParentHeader(StringBuilder sb, Id spanId) {
         spanId.writeAsHex(sb);
         sb.append('-');
         HexUtils.writeByteAsHex(flags, sb);
+    }
 
-        // todo: traceparent header- apply agreed binary format
-        for (int i = 0; i < sb.length(); i++) {
-            outgoingBinaryHeader[i] = (byte) sb.charAt(i);
+    /**
+     * Returns a binary representation of the {@code traceparent} header for downstream services.
+     * <p>
+     * NOTE: CALLER MAY NOT KEEP A REFERENCE TO THE RETURNED BYTE ARRAY, AS IT BELONGS TO THE TRACE CONTEXT INSTANCE
+     */
+    public byte[] getOutgoingTraceParentBinaryHeader() {
```

Comment:
The use of binary headers is relatively rare but adding the `outgoingBinaryHeader` variable increases the memory consumption of each object in the pool. 

Any chance to use write the value to thread-local-buffered byte arrays? Caching in `outgoingBinaryHeader` does not seem to have much benefit, as I don't see how the same TraceContext would be injected multiple times in its binary form.

```
public void fillOutgoingTraceParentBinaryHeader(byte[] header) {
```

Reply:
> The use of binary headers is relatively rare but adding the outgoingBinaryHeader variable increases the memory consumption of each object in the pool.

It's 29 bytes. What we can do is lazily create the array only when used for the first time. 

> Any chance to use write the value to thread-local-buffered byte arrays?

See the answer above- in Kafka another thread ends the span. It is possible that the thread sending the record will handle the send of another one before completion.